client/layout/masterbar/checkout.tsx

```diff
@@ -55,14 +55,23 @@ const CheckoutMasterbar: FunctionComponent< Props > = ( {
 				clearSignupDestinationCookie();
 			}
 
+			if ( searchParams.has( 'cancel_to' ) ) {
+				const cancelPath = searchParams.get( 'cancel_to' ) ?? '';
+				// Only allow redirecting to relative paths.
+				if ( cancelPath.match( /^\/[^/]?/ ) ) {
+					page( cancelPath );
+					return;
+				}
+			}
+
 			// Some places that open checkout (eg: purchase page renewals) return the
 			// user there after checkout by putting the previous page's path in the
 			// `redirect_to` query param. When leaving checkout via the close button,
 			// we probably want to return to that location also.
 			if ( searchParams.has( 'redirect_to' ) ) {
 				const redirectPath = searchParams.get( 'redirect_to' ) ?? '';
 				// Only allow redirecting to relative paths.
-				if ( redirectPath.startsWith( '/' ) ) {
+				if ( redirectPath.match( /^\/[^/]?/ ) ) {
```

Comment:
This code is being moved to a stand-alone function in https://github.com/Automattic/wp-calypso/pull/59226 so you'll probably need a rebase after that merges, but the good news is that it will affect all three places that deal with closing checkout!

Reply:
Thanks @sirbrillig. I've just rebased this after #59226 was merged. Everything seems ok still. Would you mind giving it a quick look?